### PR TITLE
0d element

### PR DESCRIPTION
**Pull request by chleh**
This is the reincarnation of Dima's PR #846. The main work has been done by him. I helped fixing some compilation errors.

This PR adds a point element to the mesh lib, suitable shape function, and makes it available for the local assemblers.

The reason behind this element is that the boundary of a line mesh consists of point elements and to implement Neumann boundary conditions, for example, integration over such elements must be performed.

The original PR suffered from compilation errors because of static allocation of zero-sized arrays (by Eigen::Matrix). This issue has been treated by always using dynamically sized Eigen matrices for zero size. Dynamic array allocation of zero size is at least legal (cf. http://stackoverflow.com/a/1087066). This solution is not perfect, but a complete solution would require major rewrites, like a distinction between dimensions starting on the level of class `ShapeMatrices`.

~~The test data will be moved to the right place upon merging this PR.~~

### CODE REVIEW

**Comment by chleh** [comment]
Failed test: `Cannot open include file: 'tclap/CmdLine.h': No such file or directory`
Is it my fault, or a problem with Jenkins?

**Comment by bilke** [comment]
Jenkins test this please

**Comment by bilke** [comment]
Jenkins builds passed. It failed because of not updating submodules due to GitHub not responding.

**Comment by endJunction** [comment]
:fast_forward:
Thanks Christoph for pushing this through!

**Comment by endJunction** [comment]
@chleh I have moved the line tests into external data and added ctests for the line. Please take my [branch](https://github.com/endJunction/ogs/tree/0dElement) instead of this one, which includes also the changes from this branch.

**Comment by chleh** [comment]
@bilke: could you please have a look at the Mac PR test?

**Comment by bilke** [comment]
Jenkins test this please.

**Comment by endJunction** [comment]
@bilke Thanks! Mac is happy again.

**Comment by endJunction** [comment]
:fast_forward:

File: MeshLib/Elements/PointRule1.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+	/// Constant: Local node index table for edge
+	static const unsigned edge_nodes[1][1];
+
+	/// Edge rule
+	typedef NoEdgeReturn EdgeReturn;
+
+	/// Checks if a point is inside the element.
+	/// \param pnt a 3D MathLib::Point3d object
+	/// \param eps tolerance for numerical algorithm used or computing the
```
Comment (TomFischer) [inline MeshLib/Elements/PointRule1.h]: - Maybe it is worth to mention that it is tested if the **square** distance it less than the given `eps`?
- spelling "... used **f**or computing ... "

File: MeshGeoToolsLib/BoundaryElementsAtPoint.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+#include "MeshGeoToolsLib/MeshNodeSearcher.h"
+
+namespace MeshGeoToolsLib
+{
+BoundaryElementsAtPoint::BoundaryElementsAtPoint(
+    MeshLib::Mesh const &mesh, MeshNodeSearcher &mshNodeSearcher,
+    GeoLib::Point const &point)
+    : _mesh(mesh), _point(point)
+{
+	auto const node_ids = mshNodeSearcher.getMeshNodeIDs(_point);
```
Comment (TomFischer) [inline MeshGeoToolsLib/BoundaryElementsAtPoint.cpp]: Why not `auto const & node_ids`?

**Pull request merged**